jitstats.lua

```diff
@@ -0,0 +1,462 @@
+    else
+        print(string.format(fmt, ...))
+    end
+end
+
+local debugprint = function () end
+
+-- Stolen from dump.lua
+local function fmtfunc(func, pc)
+    local fi = funcinfo(func, pc)
+    if fi.loc then
+        return fi.loc
+    elseif fi.ffid then
+        return vmdef.ffnames[fi.ffid]
+    elseif fi.addr then
+        return ("C:%x"):format(fi.addr)
+    else
+        return "(?)"
+    end
+end
+
+-- Format trace error message.  Stolen from dump.lua
+local function fmterr(err, info)
+    if type(err) == "number" then
+        if type(info) == "function" then 
+            info = fmtfunc(info) 
+        end
+        err = vmdef.traceerr[err]:format(info)
+    end
+    return err
+end
+
+local jitstats = {
+    isrunning = false,
+    debug = false,
+    stats = {}
+}
+
+local stats = jitstats.stats
+
+local vmevent = {}
+
+local curr_parentid, curr_stitched
+local start_pc, start_func
+-- Trace started
+function vmevent.start(tr, func, pc, parentid, exitnum)
+    curr_parentid = parentid
+    start_func = func
+    start_pc = pc
+    curr_stitched = parentid and parentid > 0 and exitnum == -1
+
+    debugprint("START(%d): parent = %d", tr, parentid or -1)
+    
+    stats.starts = stats.starts + 1
+    if curr_stitched then
+        stats.stitch_starts = stats.stitch_starts + 1
+    elseif parentid then
+        stats.side_starts = stats.side_starts + 1
+    end 
+end
+
+--Trace Stopped
+function vmevent.stop(tr)
+    local info = jutil.traceinfo(tr)
+     
+    debugprint("STOP(%d): link = %s", tr, info.linktype)
+    
+    stats.completed = stats.completed + 1
+    
+    if curr_stitched then
+        stats.stitch_completed = stats.stitch_completed + 1
+    elseif curr_parentid then
+        stats.side_completed= stats.side_completed + 1
+    end
+
+    stats.linktypes[info.linktype] = (stats.linktypes[info.linktype] or 0) + 1
+end
+
+--Trace Aborted
+function vmevent.abort(tr, func, pc, code, errinfo)
+    stats.aborts = stats.aborts + 1
+    
+    local reason
+    if jitstats.debug then
+        reason = fmterr(code, errinfo)
+        reason = reason:gsub("bytecode (%d+)", function(c)
+            c = tonumber(c)
+            assert(bcnames[c], c)
+            return "bytecode "..bcnames[c]
+        end)
+    end
+
+    if curr_stitched then
+        stats.stitch_aborts = stats.stitch_aborts + 1
+    elseif curr_parentid then
+        debugprint("ABORT(%d): parent = %d, %s", tr, curr_parentid, reason)
+        stats.side_aborts = stats.side_aborts + 1
+    else
+        debugprint("ABORT(%d): %s", tr, reason)
+    end
+    
+    local bc = funcbc(start_func, start_pc)
+    local bcname = bc and bcnames[band(bc, 0xff)]
+    if bcname == "IFUNCF" then
+        print"blacklist"
+        stats.func_blacklisted = stats.func_blacklisted + 1
+    elseif bcname == "ILOOP" or bcname == "IFORL" or bcname == "IITERL" then
+        stats.loop_blacklisted = stats.loop_blacklisted + 1
+    end
+
+    -- We sometimes don't can't find bc
```

Comment:
I can't parse the sentence: don't or can't?

Reply:
fixed